lib/pos_snapshot_entries.go

```diff
@@ -688,188 +688,169 @@ type SnapshotStakeMapKey struct {
+			)),
 		)
 	}
 
-	// Retrieve top N SnapshotStakeEntry keys by stake amount.
-	key := DBKeyForSnapshotStakeToRewardAtEpochNumber(snapshotAtEpochNumber)
-	keysFound, _, err := EnumerateKeysForPrefixWithLimitOffsetOrder(
-		handle, key, int(limit), nil, true, snapshotStakeKeysToSkip,
+	// Retrieve the snapshot StakeEntries from the DB.
+	prefix := DBKeyForSnapshotStakeToRewardAtEpochNumber(snapshotAtEpochNumber)
+	_, valsFound, err := EnumerateKeysForPrefixWithLimitOffsetOrder(
+		handle, prefix, int(limit), nil, true, snapshotStakeDBKeysToSkip,
 	)
 	if err != nil {
-		return nil, errors.Wrapf(err, "DBGetSnapshotStakesToRewardByStakeAmount:"+
+		return nil, errors.Wrapf(err, "DBGetSnapshotStakesToReward:"+
 			" problem retrieving top stakes: ")
 	}
 
-	// For each key found, parse the SnapshotStakeEntry from the key.
-	for _, keyFound := range keysFound {
-		snapshotStakeEntry, err := DecodeSnapshotStakeFromDBKey(keyFound)
+	// Decode StakeEntries from bytes.
+	var stakeEntries []*StakeEntry
+	for _, stakeEntryBytes := range valsFound {
+		rr := bytes.NewReader(stakeEntryBytes)
+		stakeEntry, err := DecodeDeSoEncoder(&StakeEntry{}, rr)
 		if err != nil {
-			return nil, errors.Wrapf(err, "DBGetSnapshotStakesToRewardByStakeAmount:"+
-				" problem reading SnapshotStakeEntry: ")
+			return nil, errors.Wrapf(err, "DBGetSnapshotStakesToReward: problem decoding StakeEntry: ")
 		}
-
-		snapshotStakeEntries = append(snapshotStakeEntries, snapshotStakeEntry)
+		stakeEntries = append(stakeEntries, stakeEntry)
 	}
-
-	return snapshotStakeEntries, nil
+	return stakeEntries, nil
 }
 
 func (bav *UtxoView) _flushSnapshotStakesToRewardToDbWithTxn(txn *badger.Txn, blockHeight uint64) error {
-	// Note: the typical DELETE -> PUT pattern we follow in other _flush functions to update re-indexed keys
-	// will not work here because the SnapshotStakeToRewardByStakeAmount index has all of the information about
-	// the snapshot stake entry in the key. We would need to know the DB stakerPKID, validatorPKID, and stakeAmount
-	// values for snapshotStakeEntry in order to delete it from the DB. If we want to follow the DELETE -> PUT, we'll
-	// need to seek through ALL keys in the index and delete them first. This is not ideal.
-	//
-	// Omitting the DELETE here is fine in practice because the snapshotStakeEntries that are written when committing
-	// the final block of an epoch will always be identical, and are never overwritten.
-
-	for mapKey, snapshotStakeEntry := range bav.SnapshotStakesToReward {
-		if snapshotStakeEntry == nil {
+	// Delete all snapshot StakeEntries in the UtxoView map.
+	for mapKeyIter, stakeEntryIter := range bav.SnapshotStakesToReward {
+		// Make a copy of the iterators since we make references to them below.
+		mapKey := mapKeyIter
+		stakeEntry := *stakeEntryIter
+
+		// Sanity-check that the entry matches the map key.
+		// TODO: Do we really need this check? We should be fine trusting the contents of the view without it.
```

Comment:
it's a little silly that we have it everywhere, but let's keep it to be consistent w/ all flush patterns.